bin/markcop

```diff
@@ -41,155 +40,145 @@ cat <<'EOF'
-      if [[ ! -z "$link_dest_file" ]]; then
-        file_to_check=$linked_path
+  return 0
+}
+
+function missing_link {
+  file_name="$1"
+  link_line_regex='[0-9]+:\[.*\]\(.*?(?=\))'
+  for line in $(echo "$2" | grep -oP $link_line_regex); do
+    line_num=$(echo $line | cut -f1)
+    dir=$(dirname "$file_name")
+    link_dest=$(echo $line | grep -oP '\[.*\]\(\K.*')
+    broken_link=true
+
+    # Only check if the link is not a URL and if it's not empty
+    if [[ ( ! $link_dest =~ $URL_REGEX ) ]]; then
+      link_dest_file=$(echo $link_dest | sed "s/#.*$//") # 'file.md' <- 'file.md#heading'
+      link_dest_heading=$(echo $link_dest | grep -o '#.*$') # '#heading' <- 'file.md#heading'
+      linked_path=$(readlink -f "$dir/$link_dest_file") # '/home/user/proj/file.md'
+
+      # If there's no link_dest_heading (the '#heading' part of 'file.md#heading')
+      if [[ -z "$link_dest_heading" ]]; then
+        # Check if the linked path is either an existing file or directory
+        if [[ ( -f "$linked_path" ) || ( -d "$linked_path" ) ]]; then
+          broken_link=false
+        fi
+      else
+        file_to_check=$file_name
+
+        # If there's a filename in the link (the 'file.md' part of 'file.md#heading')
+        if [[ ! -z "$link_dest_file" ]]; then
+          file_to_check=$linked_path
+        fi
+
+        # '#my-heading' -> 'my heading'
+        stripped_heading=$(echo $link_dest_heading | sed -e 's/#//' -e 's/-/ /g')
+
+        for match in $(
+                      cat $file_to_check |
+                        grep -i "^#\+" | # only headings
+                        sed 's/-/ /g' | # replace dashes with spaces
+                        tr -d '[:punct:]' | # remove all punctuation
+                        sed -e 's/^ \+//' -e 's/ \+$//' -e 's/  \+/ /g' | # strip extra spaces
+                        grep -io "^$stripped_heading\$"
+                    ); do
+          broken_link=false
+        done
       fi
-
-      for match in $(
-          cat $file_to_check |
-            grep -i "^#\+" | # only headings
-            sed 's/-/ /g' | # replace dashes with spaces
-            tr -d '[:punct:]' | # remove all punctuation
-            sed -e 's/^ \+//' -e 's/ \+$//' -e 's/  \+/ /g' | # strip extra spaces
-            grep -io "^$stripped_heading\$"
-        ); do
-        broken_link=false
-      done
+    else # if the link is a URL
+      broken_link=false
     fi
-  else # if the link is a URL
-    broken_link=false
-  fi
 
-  if [ $broken_link = true ]; then
-    missing_link=true
-    HAS_ERROR=true
-    printf "${RED}x${NC}"
-    errors="${errors}\n${file}:${line_num} '${link_dest}' is not a valid link!"
-  else
-    printf "${GREEN}.${NC}"
-  fi
+    if [ $broken_link = true ]; then
+      errors="${errors}\n${file}:${line_num} '${link_dest}' is not a valid link!"
+    fi
+  done
+}
+
+for file in $(git ls-files | grep '\.md$'); do
+  content=$(grep -n "" $file | perl -0777 -pe 's/[0-9]+\:```.*?```//gs')
```

Comment:
to keep it non-greedy:

```
1. ```
2. code
3. ```
4. text
5. ```
6. code
7. ```
```
The `?` ensures line 4 isn't selected